atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/cache/LockWatchValueCacheImpl.java

```diff
@@ -0,0 +1,168 @@
+import com.palantir.lock.watch.TransactionsLockWatchUpdate;
+import java.util.Comparator;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * The operations on this class *must* be synchronised to protect against concurrent modification of the cache.
+ */
+public final class LockWatchValueCacheImpl implements LockWatchValueCache {
+    private final LockWatchEventCache eventCache;
+    private final ValueStore valueStore;
+    private final SnapshotStore snapshotStore;
+
+    private volatile Optional<LockWatchVersion> currentVersion = Optional.empty();
+
+    public LockWatchValueCacheImpl(LockWatchEventCache eventCache) {
+        this.eventCache = eventCache;
+        this.valueStore = new ValueStoreImpl();
+        this.snapshotStore = new SnapshotStoreImpl();
+    }
+
+    @Override
+    public synchronized void processStartTransactions(Set<Long> startTimestamps) {
+        TransactionsLockWatchUpdate updateForTransactions =
+                eventCache.getUpdateForTransactions(startTimestamps, currentVersion);
+        updateCurrentVersion(updateForTransactions);
+        updateStores(updateForTransactions);
+    }
+
+    // TODO(jshah): This needs to be *very* carefully wired to ensure that the synchronised aspect here is not an
+    //  issue. Chances are that this may need to be re-jigged to take a batch, and be connected to the batched commit
+    //  timestamp call.
+    @Override
+    public synchronized void updateCache(TransactionDigest digest, long startTs) {
+        try {
+            CommitUpdate commitUpdate = eventCache.getCommitUpdate(startTs);
+            commitUpdate.accept(new Visitor<Void>() {
+                @Override
+                public Void invalidateAll() {
+                    // If this is an election, we should throw. If not and we are not a serialisable transaction, we are
+                    // *technically* ok to go on here. However, we cannot determine which of the two we are in, so we
+                    // throw anyway.
+                    throw new TransactionLockWatchFailedException("A Timelock leader election has occurred between the"
+                        + " start and commit time of this transaction, and thus all events have been invalidated. This"
+                        + " transaction will be retried");
+                }
+
+                @Override
+                public Void invalidateSome(Set<LockDescriptor> invalidatedLocks) {
+                    Set<CellReference> invalidatedCells = invalidatedLocks.stream()
+                            .flatMap(LockWatchValueCacheImpl::extractTableAndCell)
+                            .collect(Collectors.toSet());
+                    KeyedStream.stream(digest.loadedValues())
+                            .filterKeys(cellReference -> !invalidatedCells.contains(cellReference))
+                            .forEach(valueStore::putValue);
```

Comment:
So this again seems like it's ripe for race conditions.

What if this is processed first, and then we process a start transactions request that actually saw fewer events than us? Seems that we should also update the `currentVersion` here to protect against that.

Also, if any of the puts throws, we ignore the rest which is a bit wasteful but probably fine.

Reply:
This is actually not an issue since this does _not_ show us all events. The commit update _only_ shows you lock events, and so you cannot even apply the events at commit time.

It doesn't matter if you saw 1000 events in this commit update, then the next start transactions request only showed you one, as those events would _just_ filter out the reads from that transaction (in other words, it would just be more pessimistic). It doesn't affect the correctness or state of the cache.

Put differently, if we decided to randomly not flush X reads to the central cache, it wouldn't make it incorrect, just inefficient. This is the same, except it's not random and hopefully not inefficient :P

Reply:
So, because the methods are synchronized, and presumably the method that calls `processStartTransactions` is also synchronised (since transaction starter), it is guaranteed that any transaction calling it must have started after the transaction calling update cache on commit. Therefore, anything we put in will have been visible by that other transaction since it was also visible at the start of our transaction. If we remove something, it may have been visible, but whatever

Reply:
Yes - this is doubly guaranteed by the sequence. Either the new transactions have a later or equal sequence, or they don't (and we throw).